### DISCUSSION TITLE
Add REST api in front of DGS in same spring boot project

### DISCUSSION THREAD

**Discussion by aisensiy** [Netflix/dgs-framework#998] [Q&A]
I found that graphql is really convenient to generate complex query result. For example a schema like this:

```graphql
type Query {
  article(id: String!): Article
}

type User {
  username: String!
  email: String!
}

type Article {
  id: String!
  title: String!
  author: User!
  content: String!
  createdAt: DateTime!
  author: User!
  comments: [Comment!]!
}

type Comment {
  id: String!
  article: Article!
  author: User!
  content: String!
  createdAt: DateTime!
}
```

If I want query an article including its author, the query can be like this:

```graphql
query {
  article(id: "aabb") {
    id
    title
    content
    author {
      username
      email
    }
  }
}
```

My question is that it is possible to create a RESTful API and use the dgs as query service in the same spring boot project like below:

```java
@RestController
public class ArticleApi {
  private DgsService dgsService;

  @GetMapping("/articles/{articleId}")
  public ArticleResponse getArticleById(@PathParameter("articleId") String articleId) {
    return ArticleResponse.build(dgsService.query("<the graphql query>");
  } 
}
```

**Comment by oonxt** [comment]
try this:
https://netflix.github.io/dgs/advanced/java-client/

**Reply by aisensiy** [reply]
Ok. This is not unaccpetable and thanks for your idea. But is there some solution without the http loop?

**Reply by paulbakker** [reply]
You can just use the `DgsQueryExecutor` (which you can inject as a bean), similar to how you would for [testing](https://netflix.github.io/dgs/query-execution-testing/#example). This way you skip the "web" layer in the framework and just use the query execution.

**Reply by owengalenjones** [reply]
@paulbakker I'm attempting to follow this advice to put DGS in a Spring app that uses JAX-RS as the controllers. What you have described above is working so far, but I'm not sure how to handle the actual request body String.

```
 @POST
  public Object graphql(@Language("GraphQL") @RequestBody String body) {
    var result =  executor.execute(body);
    var data = result.getData();
    return data;
  }
```

This approach doesn't work as the `body` doesn't appear to be formatted in a way the executor likes. I receive the following error:

```
Query did not parse : '{"query":"query Library {library {library_id}}","operationName":"Library","variables":{}}'
```

This query was created using a GraphQL client and is valid GraphQL and conforms to the schema.

When I tried creating the request manually in the code this worked for me:

```
  @POST
  public Object graphql(@Language("GraphQL") @RequestBody String body) {
    var s = "{ library {library_id}}";
    var result =  executor.execute(s);
    var data = result.getData();
    return data;
  }
```

Any ideas what I would need to do to turn the `String` into a valid query for the executor?

Thank you!

**Reply by srinivasankavitha** [reply]
You'll need to do something similar to how we handle this in the DGS framework: https://github.com/Netflix/dgs-framework/blob/master/graphql-dgs-spring-webmvc/src/main/kotlin/com/netflix/graphql/dgs/mvc/DgsRestController.kt#L100

**Reply by owengalenjones** [reply]
Thanks, I think I got it!

**Reply by srikanthrks** [reply]
@owengalenjones were you able to execute the graphql query in your REST method implementation. Looking for a similar way to call the method annotated with `@DgsQuery` (`getArticle`) from within `@GetMapping("/articles/{articleId}")
  public ArticleResponse getArticleById()`

```@DgsComponent
public class ArticleDataFetcher {
  @DgsQuery
  public Article getArticle(@InputArgument String id) {
    if(id.isEmpty()){
      throw new IllegalArgumentException("id is required");
    }
    return new Article();
  }
```

**Reply by owengalenjones** [reply]
@srikanthrks: If you need it in a RestMVC `@GetMapping` I have not personally tested it, but the DGS / graphql-java executor is accessible and you can probably call it from wherever:

```
public class GetMappingDGSAdapter {

  private final DgsQueryExecutor executor;

  @GetMapping("/articles/{articleId}")
  public ArticleResponse getArticleById() {

    var body = new GraphQLRequestBody(...)

      return executor.execute(
          body.getQuery(),
          body.getVariables(),
          Map.of(),
          null,
          body.getOperationName(),
          null
      );
  }

  ...

  public static class GraphQLRequestBody {

    private String query;
    private String operationName;
    private Map<String, Object> variables = new LinkedHashMap<>();
  }
```

Once the `executor` is invoked it should find the `@DgsQuery` annotated method and invoke it.

**Reply by srikanthrks** [reply]
@owengalenjones Thanks for the suggestion. Will try this out
